### PR TITLE
gcc8: init at 8.1.0

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Motivation for this change

GCC 8.1 was just released.

###### Things done

- ~~This is still compiling on my machine. I see no reason why it wouldn't finish, but I wanted to get the WIP PR out.~~ Okay; it compiled. Ship it? 
- This does _not_ set GCC8 as the default version. 
- For good measure, this derivation is mostly copied from GCC7

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by dtzWill** [comment]
I vote getting this in the repository first, with separate PR for changing the default (and dealing with any breakage/fixes needed).

**Comment by Synthetica9** [comment]
Sounds fair enough

File: pkgs/development/compilers/gcc/8/default.nix [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+assert x11Support -> (filter (x: x == null) ([ gtk2 libart_lgpl ] ++ xlibs)) == [];
+
+stdenv.mkDerivation ({
+  name = "${name}${if stripped then "" else "-debug"}-${version}" + crossNameAddon;
+
+  builder = ../builder.sh;
+
+  src = fetchurl {
+    url = "mirror://gcc/releases/gcc-${version}/gcc-${version}.tar.xz";
+    /* sha256 = "0p71bij6bfhzyrs8676a8jmpjsfz392s2rg862sdnsk30jpacb43"; */
```
[the remark this answers is not in the document]
Comment (Synthetica9) [reply pkgs/development/compilers/gcc/8/default.nix]: Whoops, :heavy_check_mark:

**Comment by dtzWill** [comment]
Consider adding these commits:
* add cc-wrapper variants for gcc8: https://github.com/dtzWill/nixpkgs/commit/9ecd2e2e04df3ee3f90a46a42685f86d23d38829
* drop self-tests patch as it doesn't apply, we'll see how things work!
https://github.com/dtzWill/nixpkgs/commit/ec4a46098c3fcd9e6abb78791cff16f66e75fdb0

(for testing purposes I made a branch that contains this as well as setting gcc8 as default: 
https://github.com/dtzWill/nixpkgs/commits/experimental/gcc8)

**Comment by Synthetica9** [comment]
@dtzWill The `drop-self-test` patch _was_ available in 7.3.0 (https://git.busybox.net/buildroot/plain/package/gcc/7.3.0), so I don't we should remove that yet (unless it was somehow confirmed by the GCC dev team that it won't be needed)

**Comment by dtzWill** [comment]
> @dtzWill The drop-self-test patch was available in 7.3.0 (https://git.busybox.net/buildroot/plain/package/gcc/7.3.0), so I don't we should remove that yet (unless it was somehow confirmed by the GCC dev team that it won't be needed)

Well since the patch wasn't issued by the GCC dev team I'm not sure why they're the authority to ask :).

But leaving it commented-out works for me-- the real answer is to get some builders testing cross to see if it is needed or not.  That can be handled separately, sorry for piling on :).

**Comment by Synthetica9** [comment]
@dtzWill Nah, it's fine. I'll change the hash to a placeholder value and make the url dependent on the version, but leave it commented. 

How's the gcc7->gcc8 switch looking with regards to broken packages?

**Comment by dtzWill** [comment]
> How's the gcc7->gcc8 switch looking with regards to broken packages?

Not sure--and I won't be able to provide a good answer for this, someone would need to setup a new jobset on Hydra.  But that can be done in the follow-up PR that changes  the default?

I can confirm that with gcc8 as default we're at least able to build
* x86_64-linux: through cc-test-wrapper
* x86_64-musl: through test-wrapper
* x86_64 -> aarch64-musl:stdenv.cc (no self-tests patch!)
* x86_64 -> aarch64-glibc: errror building glibc itself D:

**Comment by dtzWill** [comment]
LGTM, borg is probably having same problem I was having with mirrors (ended up manually using `nix-prefetch-url`).  We can at least make it available for testing with this!

**Review by dtzWill — APPROVED** [review]

File: pkgs/development/compilers/gcc/8/default.nix [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+, libpthread ? null, libpthreadCross ? null  # required for GNU/Hurd
+, stripped ? true
+, gnused ? null
+, cloog # unused; just for compat with gcc4, as we override the parameter on some places
+, darwin ? null
+, buildPlatform, hostPlatform, targetPlatform
+, buildPackages
+}:
+
+assert langJava     -> zip != null && unzip != null
```
Comment (gebner) [inline pkgs/development/compilers/gcc/8/default.nix]: I thought upstream had removed java support in gcc 6?

Comment (Synthetica9) [reply pkgs/development/compilers/gcc/8/default.nix]: Oh, yeah, right you are. Should the arguments stay in place and go unused (apart from maybe an `assert not langJava`), or should I remove them altogether (https://github.com/NixOS/nixpkgs/blob/ba52ae50488de85a9cf60a3a04f1c9ca7122ec74/pkgs/development/compilers/gcc/7/default.nix#L30 seems to suggest that the first option would be more conventional.)

Comment (gebner) [reply pkgs/development/compilers/gcc/8/default.nix]: I think we should remove them completely.  AFAICT the only place where we override enableJava et al. is in the gcj6 attribute (which uses gcc6).

**Comment by Synthetica9** [comment]
@gebner I removed all mentions of Java. I'll also create a PR to do the same for GCC7.

**Comment by shlevy** [comment]
Unable to test at the moment, but IIRC gcc 8 contains the RISC-V stuff already

**Comment by Synthetica9** [comment]
@shlevy I `ctrl-f`'d the [release notes](https://gcc.gnu.org/gcc-8/changes.html) for RISC, but it isn't mentioned AFAICS.

**Comment by shlevy** [comment]
Just checked, both patches are on the 8.1.0 tag on subversion.

File: pkgs/development/compilers/gcc/8/default.nix [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+, profiledCompiler ? false
+, staticCompiler ? false
+, enableShared ? true
+, texinfo ? null
+, perl ? null # optional, for texi2pod (then pod2man)
+, gmp, mpfr, libmpc, gettext, which
+, libelf                      # optional, for link-time optimizations (LTO)
+, isl ? null # optional, for the Graphite optimization framework.
+, zlib ? null
+, pkgconfig ? null
```
Comment (gebner) [inline pkgs/development/compilers/gcc/8/default.nix]: With java support removed, all of these options (`pkgconfig` until `x11Support`) are now unused as well.

Comment (Synthetica9) [reply pkgs/development/compilers/gcc/8/default.nix]: _grabs hatchet_

**Comment by dtzWill** [comment]
This is ready to be merged, just waiting to confirm doesn't trigger any mass-rebuilds.

**Comment by dtzWill** [comment]
LGTM, apologies if I missed something.  Thanks @Synthetica9!

**Comment by Synthetica9** [comment]
@dtzWill: I think the next step would be a PR to use GCC8 as the default GCC version? Would it be sufficient to just `git cherry-pick` https://github.com/dtzWill/nixpkgs/commit/2e7772a0494e2c7a6bcc0c1766e6cf40137585ea from master and throw that up for PR?

**Comment by dtzWill** [comment]
@Synthetica9 I think so! There definitely is some breakage but we can address that in that PR :).

(and if my commit isn't sufficient for whatever reason that can be discussed/reviewed there as well)

**Comment by matthewbauer** [comment]
Hi,

gcc8 is failing on x86_64-darwin:

Job: https://hydra.nixos.org/job/nixpkgs/trunk/gcc8.x86_64-darwin

Not sure what's going on. Clang gives this error:

```
clang-5.0: error: unsupported option '--gstabs'
```

**Comment by Synthetica9** [comment]
@matthewbauer I think it would be best to open a separate issue for that; I don't have access to a macos box.

**Pull request merged**